src/OpenTelemetry.Instrumentation.StackExchangeRedis/README.md

```diff
@@ -71,6 +71,68 @@ the `ConfigureServices` of your `Startup` class. Refer to documentation for
 For an ASP.NET application, adding instrumentation is typically done in the
 `Global.asax.cs`. Refer to documentation for [OpenTelemetry.Instrumentation.AspNet](../OpenTelemetry.Instrumentation.AspNet/README.md).
 
+## Specify the Redis connection
+
+The following sections cover different ways to specify the `StackExchange.Redis`
+connection(s) that will be instrumented and captured by OpenTelemetry.
+
+### Pass a Redis connection when calling AddRedisInstrumentation
+
+The simplest thing to do is pass a created connection to the
+`AddRedisInstrumentation` extension method:
+
+```csharp
+using var connection = ConnectionMultiplexer.Connect("localhost:6379");
+
+using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+    .AddRedisInstrumentation(connection)
+    .Build();
+```
+
+Whatever connection is specified will be collected by OpenTelemetry.
+
+### Use the application IServiceProvider
+
+Users using the `OpenTelemetry.Extensions.Hosting` package may prefer to manage
+the Redis connection via the application `IServiceCollection`. To support this
+scenario, if a connection is not passed to the `AddRedisInstrumentation`
+extension manually one will be resolved one using the `IServiceProvider`:
+
+```csharp
+appBuilder.Services.AddSingleton<IConnectionMultiplexer>(
+    sp => MyRedisConnectionHelper.CreateConnection(sp));
+
+appBuilder.Services
+    .AddOpenTelemetry()
+    .WithTracing(tracing => tracing.AddRedisInstrumentation());
+```
+
+Whatever connection is found in the `IServiceProvider` will be collected by
+OpenTelemetry.
+
+### Interact with StackExchangeRedisInstrumentation directly
+
+For full control of the Redis connection(s) being instrumented the
+`ConfigureRedisInstrumentation` extension is provided to expose the
+`StackExchangeRedisInstrumentation` class directly:
+
+```csharp
+StackExchangeRedisInstrumentation redisInstrumentation = null;
+
+using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+    .AddRedisInstrumentation()
+    .ConfigureRedisInstrumentation(instrumentation => redisInstrumentation = instrumentation)
+    .Build();
+
+using var connection1 = ConnectionMultiplexer.Connect("localhost:6379");
+redisInstrumentation.AddConnection(connection1);
```

Comment:
I don't have strong feelings against this pattern, but it does seem a bit awkward to me. Given that we add instrumentation to a tracer provider it might feel more natural if we had a way to ask the tracer provider for a handle to the instrumentation. Something like

```csharp
using var tracerProvider = Sdk.CreateTracerProviderBuilder()
    .AddRedisInstrumentation()
    .Build();

using var connection1 = ConnectionMultiplexer.Connect("localhost:6379");
tracerProvider.GetInstrumentation<StackExchangeRedisInstrumentation>().AddConnection(connection1);
```

But I also don't feel strongly about adding a `GetInstrumentation` method to TracerProvider either since this Redis instrumentation is the only instrumentation where this seems to make sense today - and of course most of our instrumentation classes are not public anyways.

This is just food for thought, at the end of the day I understand this is the advanced scenario.

Reply:
It is a very interesting idea. We had the need somewhere (PrometheusExporter?) to find the [registered exporter](https://github.com/open-telemetry/opentelemetry-dotnet/blob/1510b76634f4fdfbf79cd73e4b71aaf474e40a46/src/OpenTelemetry/Metrics/MeterProviderExtensions.cs#L114) so this type of thing does seem to have some utility. But I'm not sure we have seen _enough_ need to justify a public API.

This particular API (`ConfigureRedisInstrumentation`) isn't actually needed at all. Users _could_ do this...

```csharp
        var builder = Sdk.CreateTracerProviderBuilder()
            .AddRedisInstrumentation();

        StackExchangeRedisInstrumentation redisInstrumentation = null;

        ((IDeferredTracerProviderBuilder)builder).Configure((sp, builder) =>
            redisInstrumentation = sp.GetRequiredService<StackExchangeRedisInstrumentation>());

        using var tracerProvider = builder.Build();
```

I just thought it was so nasty we should have a helper to clean it up 🤣

Using `IServiceCollection` style it is much nicer:

```csharp
appBuilder.Services.AddOpenTelemetry().WithTracing(builder => builder.AddRedisInstrumentation());

var app = appBuilder.Build();
var instrumentation = app.Services.GetRequiredService<StackExchangeRedisInstrumentation>();
```

The reason the first one is so nasty is in the "Sdk.Create" style we don't expose the `IServiceProvider` directly so it is kind of hard to get at. Might also be a useful public API?